src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -11,13 +11,64 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
+    /// <summary>
+    /// A builder for lazy instances of TypeSymbolWithAnnotations.
+    /// </summary>
+    [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+    internal struct TypeSymbolWithAnnotationsBuilder
+    {
+        private TypeSymbol _defaultType;
+        private INonNullTypesContext _nonNullTypesContext;
+        private bool _isAnnotated;
+        private TypeSymbolWithAnnotationsImplementation _implementation;
+
+        internal TypeSymbol DefaultType => _defaultType;
+        internal INonNullTypesContext NonNullTypesContext => _nonNullTypesContext;
+        internal bool IsAnnotated => _isAnnotated;
+        internal TypeSymbolWithAnnotationsImplementation Implementation => _implementation;
+
+        internal bool IsNull => _defaultType is null;
+
+        internal bool InterlockedInitialize(TypeSymbolWithAnnotations type)
+        {
+            if ((object)_defaultType != null)
+            {
+                return false;
+            }
+            _isAnnotated = type.IsAnnotated;
+            Interlocked.CompareExchange(ref _nonNullTypesContext, type.NonNullTypesContext, null);
+            Interlocked.CompareExchange(ref _implementation, type.Implementation, null);
+            return (object)Interlocked.CompareExchange(ref _defaultType, type.DefaultType, null) == null;
+        }
+
+        internal TypeSymbolWithAnnotations ToType()
+        {
+            return new TypeSymbolWithAnnotations(DefaultType, NonNullTypesContext, IsAnnotated, Implementation);
+        }
+
+        internal string GetDebuggerDisplay() => ToType().GetDebuggerDisplay();
+    }
+
     /// <summary>
     /// A simple class that combines a single type symbol with annotations
     /// </summary>
     [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
-    internal abstract class TypeSymbolWithAnnotations : IFormattable
+    internal struct TypeSymbolWithAnnotations : IFormattable
     {
-        public sealed override string ToString() => TypeSymbol.ToString();
+        internal readonly TypeSymbol DefaultType;
```

Comment:
`DefaultType` and `Implementation` were not obvious. Maybe useful to rename and/or comment.